cmd-graph.c

```diff
@@ -71,6 +71,173 @@ static bool full_graph = false;
+	const char *name;
+	const char *header;
+	int length;
+	bool used;
+	void (*print)(struct graph_node *node);
+};
+
+static void print_duration(struct graph_node *node)
+{
+	uint64_t d = 0;
+
+	d = node->time;
+
+	print_time_unit(d);
+}
+
+static void print_addr(struct graph_node *node)
+{
+	/* uftrace records (truncated) 48-bit addresses */
+	int width = sizeof(long) == 4 ? 8 : 12;
+
+	pr_out("%*lx", width, node->addr);
+}
+
+static struct graph_field field_duration = {
+	.id      = GRAPH_F_DURATION,
+	.name    = "duration",
+	.header  = " DURATION ",
+	.length  = 10,
+	.print   = print_duration,
+	.list    = LIST_HEAD_INIT(field_duration.list),
+};
+
+static struct graph_field field_addr = {
+	.id      = GRAPH_F_ADDR,
+	.name    = "addr",
+#if __SIZEOF_LONG == 4
+	.header  = "  ADDR  ",
+	.length  = 8,
+#else
+	.header  = "   ADDRESS  ",
+	.length  = 12,
+#endif
+	.print   = print_addr,
+	.list    = LIST_HEAD_INIT(field_addr.list),
+};
+
+static void print_header(void)
+{
+	struct graph_field *field;
+
+	pr_out("#");
+	list_for_each_entry(field, &output_fields, list)
+		pr_out("%s ", field->header);
+
+	if (!list_empty(&output_fields))
+		pr_out(" ");
+
+	pr_out(" FUNCTION\n");
+}
+
+/* index of this table should be matched to graph_field_id */
+static struct graph_field *field_table[] = {
+	&field_duration,
+	&field_addr,
+};
+
+static void print_field(struct graph_node *node)
+{
+	struct graph_field *field;
+
+	if (list_empty(&output_fields))
+		return;
+
+	pr_out(" ");
+	list_for_each_entry(field, &output_fields, list) {
+		field->print(node);
+		pr_out(" ");
+	}
+	pr_out(": ");
+}
+
+static void print_empty_field(void)
+{
+	struct graph_field *field;
+
+	if (list_empty(&output_fields))
+		return;
+
+	pr_out(" ");
+	list_for_each_entry(field, &output_fields, list) {
+		pr_out("%*s", field->length, "");
+		pr_out(" ");
+	}
+	pr_out(": ");
+}
+
+static void add_field(struct graph_field *field)
+{
+	if (field->used)
+		return;
+
+	field->used = true;
+	list_add_tail(&field->list, &output_fields);
+}
+
+static void setup_field(struct opts *opts)
+{
+	struct graph_field *field;
+	unsigned i;
+	char *str, *p, *s;
+
+	/* default fields */
+	if (opts->fields == NULL) {
+		add_field(field_table[GRAPH_F_DURATION]);
+		return;
+	}
+
+	if (!strcmp(opts->fields, "none"))
+		return;
+
+	s = str = xstrdup(opts->fields);
+
+	if (*str == '+') {
+		/* prepend default fields */
+		add_field(field_table[GRAPH_F_DURATION]);
+		s++;
+	}
+
+	p = strtok(s, ",");
+	while (p) {
+		for (i = 0; i < ARRAY_SIZE(field_table); i++) {
+			field = field_table[i];
+
+			if (strcmp(field->name, p))
+				continue;
+
+			add_field(field);
+			break;
+		}
+
+		if (i == ARRAY_SIZE(field_table)) {
+			pr_out("uftrace: Unknown field name '%s'\n", p);
+			pr_out("uftrace:   Possible fields are:");
+			for (i = 0; i < ARRAY_SIZE(field_table); i++)
+				pr_out(" %s", field_table[i]->name);
+			pr_out("\n");
+			exit(1);
+		}
+
+		p = strtok(NULL, ",");
+	}
+
+	free(str);
+}
+
```

Comment:
This code is a duplication, it'd be nice if you could find a way to share the code (if possible).

Reply:
I've done it in `utils/field.[ch]`. Thanks!